ddtrace/internal/utils/retry.py

```diff
@@ -0,0 +1,62 @@
+from __future__ import absolute_import
+
+from functools import wraps
+from itertools import repeat
+import random
+import sys
+from time import sleep
+import typing as t
+
+import six
+
+
+class RetryError(Exception):
+    pass
+
+
+def retry(after, until=lambda result: result is None):
+    # type: (t.Union[int, float, t.Iterable[t.Union[int, float]]], t.Callable[[t.Any], bool]) -> t.Callable
+    def retry_decorator(f):
+        @wraps(f)
+        def retry_wrapped(*args, **kwargs):
+            after_iter = repeat(after) if isinstance(after, (int, float)) else after
+            exc_info = None
+
+            for s in after_iter:
+                try:
+                    result = f(*args, **kwargs)
+                except Exception as e:
+                    exc_info = sys.exc_info()
+                    result = e
+
+                if until(result):
+                    return result
+
+                sleep(s)
+
+            # Last chance to succeed
+            try:
+                result = f(*args, **kwargs)
+            except Exception as e:
+                exc_info = sys.exc_info()
+                result = e
+
+            if until(result):
```

Comment:
maybe we should have callback like node.js: until(result,error) so it is "type-safe" and can do if error is None - but not sure this is a the "python way"

Reply:
Nice idea, probably more suited for async-style Python though 🙂